mmv1/third_party/terraform/tpgresource/common_diff_suppress.go.erb

```diff
@@ -179,25 +179,47 @@ func TimestampDiffSuppress(format string) schema.SchemaDiffSuppressFunc {
 	}
 }
 
-// suppress diff when saved is Ipv4 and Ipv6 format while new is required a reference
-// this happens for an internal ip for Private Services Connect
+// Suppresses diff for IPv4 and IPv6 different formats.
+// It also suppress diffs if one IP and one reference are given.
 func InternalIpDiffSuppress(_, old, new string, _ *schema.ResourceData) bool {
-	olds := strings.Split(old, "/")
-	news := strings.Split(new, "/")
-
-	if len(olds) == 2 {
-		if len(news) == 2 {
-			return bytes.Equal(net.ParseIP(olds[0]), net.ParseIP(news[0])) && olds[1] == news[1]
-		} else {
-			return (net.ParseIP(olds[0]) != nil) && (net.ParseIP(new) == nil)
+	addr_netmask_old := strings.Split(old, "/")
+	addr_netmask_new := strings.Split(new, "/")
+
+	// Check if old or new are IPs (with or without netmask)
+	addr_old := func() net.IP {
+		if net.ParseIP(addr_netmask_old[0]) == nil {
+			return net.ParseIP(addr_netmask_old[0])
+		}
+		return net.ParseIP(old)
+	}()
+	addr_new := func() net.IP {
+		if net.ParseIP(addr_netmask_new[0]) == nil {
+			return net.ParseIP(addr_netmask_new[0])
 		}
+		return net.ParseIP(new)
+	}()
+
+	// If old or new are references, return true
+	if (addr_old == nil) || (addr_new == nil) {
+		return true
 	}
 
-	if (net.ParseIP(old) != nil) && (net.ParseIP(new) != nil) {
-		return bytes.Equal(net.ParseIP(old), net.ParseIP(new))
+	addr_equality := false
+	netmask_equality := false
+
+	if bytes.Equal(addr_old, addr_new) {
+		addr_equality = true
+	}
+
+	// If old and new have a netmask compare them,
+	// otherwise always assume they are the same
```

Comment:
I'm not super familiar with IPv6 - why is this a safe assumption to make? (A link to documentation explaining this would be great!)

Reply:
Unfortunately, this is not safe and it's not documented anywhere but it's exactly what we're trying to solve with this patch. Please look at the bug we're referencing ([#16400](https://github.com/hashicorp/terraform-provider-google/issues/16400)).